packages/sheets/src/commands/mutations/set-range-values.mutation.ts

```diff
@@ -123,50 +124,61 @@ export const SetRangeValuesMutation: IMutation<ISetRangeValuesMutationParams, bo
         newValues.forValue((row, col, newVal) => {
             // clear all
             if (!newVal) {
-                cellMatrix?.setValue(row, col, {});
+                cellMatrix.realDeleteValue(row, col);
             } else {
-                const oldVal = cellMatrix.getValue(row, col) || {};
+                let oldVal = cellMatrix.getValue(row, col) || {};
+                oldVal = mergeCellData(newVal, oldVal, styles);
 
-                // NOTE: we may need to take `p` into account
-                const type = getCellType(styles, newVal, oldVal);
-
-                if (newVal.f !== undefined) {
-                    oldVal.f = newVal.f;
-                }
-
-                if (newVal.si !== undefined) {
-                    oldVal.si = newVal.si;
-                }
-
-                if (newVal.p !== undefined) {
-                    oldVal.p = newVal.p;
-                }
-
-                // Set to null, clear content
-                if (newVal.v !== undefined) {
-                    oldVal.v = getCellValue(type, newVal);
-                }
-
-                if (oldVal.v !== undefined) {
-                    oldVal.t = type;
-                    // The text format may be set for the numeric cell, and cell.v needs to be converted to text
-                    oldVal.v = getCellValue(type, oldVal);
-                }
-
-                // handle style
-                if (newVal.s !== undefined) {
-                    handleStyle(styles, oldVal, newVal);
-                }
-
-                if (newVal.custom !== undefined) {
-                    // Custom will overwrite the original value
-                    oldVal.custom = newVal.custom;
+                if (Tools.isEmptyObject(oldVal)) {
+                    cellMatrix.realDeleteValue(row, col);
+                } else {
+                    cellMatrix.setValue(row, col, oldVal);
                 }
-
-                cellMatrix.setValue(row, col, Tools.removeNull(oldVal));
             }
         });
 
         return true;
     },
 };
+
+function mergeCellData(newValue: ICellData, oldValue: ICellData, styles: Styles) {
+    const overwriteCellPropertiesSet = new Set(['f', 'p', 'si', 'custom']);
+    Object.keys(newValue).forEach((key) => {
+        const cellPropertyKey = key as keyof ICellData;
+        if (overwriteCellPropertiesSet.has(cellPropertyKey)) {
+            const propertyValue = newValue[cellPropertyKey];
+            updateCellProperty(oldValue, cellPropertyKey, propertyValue);
+        } else if (cellPropertyKey === 'v') {
+            const type = getCellType(styles, newValue, oldValue);
+            if (newValue.v !== undefined) {
+                oldValue.v = getCellValue(type, newValue);
+            }
+            if (oldValue.v !== undefined) {
```

Comment:
Not equivalent, however, whether the old value requires t may also be an optimization point